flake.nix

```diff
@@ -0,0 +1,382 @@
+                pygments
+                markdown
+                mkdocs
+                pymdown-extensions
+                jinja2
+                colorama
+                regex
+                requests
+              ];
+              doCheck = false;
+            };
+          editdistpy = with pythonPkgs;
+            buildPythonPackage rec {
+              pname = "editdistpy";
+              version = "0.1.3";
+              src = fetchPypi {
+                inherit pname version;
+                hash = "sha256-s8rQcxnXn+izumv5IpPelikykX2I6cYk33tqRL3fHcw=";
+              };
+            };
+          symspellpy = with pythonPkgs;
+            buildPythonPackage rec {
+              pname = "symspellpy";
+              version = "6.7.7";
+              src = fetchPypi {
+                inherit pname version;
+                hash = "sha256-9sMVGHeAvC3TD8nKMu8Hb4m7/LKns/mjd5JvH2reAIU=";
+              };
+              buildInputs = [ setuptools ];
+              propagatedBuildInputs = [ editdistpy ];
+              format = "pyproject";
+            };
+          mkdocs-spellcheck = with pythonPkgs;
+            buildPythonPackage rec {
+              pname = "mkdocs-spellcheck";
+              version = "0.2.1";
+              src = fetchPypi {
+                inherit pname version;
+                hash = "sha256-g8neboAWGGN04EWsSBKj4oHyKVN/iKP4wANO+Ba3nI4=";
+              };
+              format = "pyproject";
+              buildInputs = [
+                pdm-pep517
+              ];
+              propagatedBuildInputs = [
+                symspellpy
+              ];
+            };
+          pythonEnv = pkgs.python310.withPackages (ps: [
+            ps.pytest
+            ps.typing-extensions
+            ps.mypy
+            ps.autopep8
+            ps.pip
+            mkdocs
+            mkdocs-material-extensions
+            mkdocs-material
+            mkdocs-spellcheck
+          ]);
+
+          mkEnvSerialize = (envKey: envValue: "export ${envKey}=${envValue};");
+          mkEnv = (envAttrs:
+            lib.concatStrings
+              (lib.mapAttrsToList
+                mkEnvSerialize
+                envAttrs)
+          );
+          mkExec = (execName: envAttrs: execArgs:
+            "${mkEnv envAttrs}${execName} ${execArgs}"
+          );
+          controllerCmd = mkExec "workflow-controller" argoConfig.controller.env argoConfig.controller.args;
+          argoServerCmd = mkExec "argo" argoConfig.argoServer.env argoConfig.argoServer.args;
+          uiCmd = mkExec "yarn" argoConfig.ui.env argoConfig.ui.args;
+        in
+        {
+          packages = {
+            ${package.name} = pkgs.buildGoModule {
+              pname = package.name;
+              inherit (package) version;
+              inherit src;
+              vendorSha256 = "sha256-1ivPt7m65B89jOTlYA21AcrW6OZeMlgpoPZd9Lw2Ohc=";
+              doCheck = false;
+            };
+
+            mockery = pkgs.buildGoModule rec {
```

Comment:
Can these be automatically updated since we have versions in other places too or somehow reuse existing Makefile?

Reply:
There is `nix flake update` but nothing to update them to exactly the version in the Makefile. 
We can't really reuse the Makefile, Nix is designed to be completely sealed off, this is really it's core advantage. Because it manages dependencies itself, it can guarantee that the build will be reproducible on any machine whether it be a Ubuntu, Fedora or even 20 years in the future.

Reply:
Maybe we can add clear, hard-to-miss comments in the Makefile regarding how to go about modifying Nix, and possibly the other way around as well?

Does it make any sense to have CI compare the two?

Reply:
Yeah I think just documentation is really the best we can do unfortunately, I think this is fine anyway if we explicitly state in the docs that this feature is experimental and not supported (and should therefore not be relied on in production etc). This is very much a solution for developers only, to get them developing as quickly as possible. 

I can't really think of how we can compare them in the CI process unfortunately.

Reply:
I'll write some script to notify me on a Makefile change, so I can get to it ASAP. I have the capacity (and get paid) to do this kind of ad-hoc work anyway, so shouldn't be a problem.